client/lib/ffi.rs

```diff
@@ -0,0 +1,676 @@
+    for bytes in slice {
+        // While null-ptr strings are usually allowed as single arguments, an array of strings
+        // required to not contain null values.
+        if bytes.is_null() {
+            return Err(Error::FFIPtrNullButRequired(arg_name));
+        }
+        vec.push(unsafe_str_arg(*bytes, arg_name)?);
+    }
+    Ok(vec)
+}
+
+/// Helper to call TryInto::try_into on a *const ptr of our rust type implementing it.
+/// This is used for
+fn unsafe_try_into<T, I>(value: *const I, field_name: &'static str) -> Result<T>
+where
+    I: Clone,
+    I: TryInto<T, Error = Error>,
+{
+    if value.is_null() {
+        Err(Error::FFIPtrNullButRequired(field_name))
+    } else {
+        let value: T = unsafe { (*value).clone().try_into()? };
+        Ok(value)
+    }
+}
+
+/// Copy the contents of `strval` to a user-provided buffer.
+///
+/// `strval` is the rust `&str` utf8 string to copy.
+/// `buf` is the caller-provided buffer to write into.
+/// `len` is the size of the buffer `buf` in bytes.
+/// - returns the number of bytes written to `buf`.
+fn copy_str_to_buf(strval: &str, buf: *mut c_uchar, len: usize) -> usize {
+    let mut_buf = unsafe { slice::from_raw_parts_mut::<u8>(buf, len) };
+    let lesser_len = len.min(strval.len());
+    let strval = strval.as_bytes();
+    mut_buf[0..lesser_len].clone_from_slice(&strval[0..lesser_len]);
+    len
+}
+
+/// Perform needed setup for the client library.
+#[no_mangle]
+pub extern "C" fn casper_setup_client() {
+    let mut runtime = RUNTIME.lock().expect("should lock");
+    // TODO: runtime opts
+    *runtime = Some(runtime::Runtime::new().expect("should create tokio runtime"));
+}
+
+/// Perform a clean shutdown of resources gathered in the client library.
+#[no_mangle]
+pub extern "C" fn casper_shutdown_client() {
+    let mut runtime = RUNTIME.lock().expect("should lock");
+    *runtime = None; // triggers drop on our runtime
+}
+
+/// Gets the last error copied to the provided buffer.
+///
+/// * `buf` is the buffer where the result will be stored.
+/// * `len` is the length of the `buf` buffer in bytes.
+/// - returns the number of bytes written to `buf`.
+#[no_mangle]
+pub extern "C" fn casper_get_last_error(buf: *mut c_uchar, len: usize) -> usize {
+    if let Some(last_err) = &*LAST_ERROR.lock().expect("should lock") {
+        let err_str = format!("{}", last_err);
+        return copy_str_to_buf(&err_str, buf, len);
+    }
+    0
+}
+
+/// Creates a `Deploy` and sends it to the network for execution.
+///
+/// See [super::put_deploy](function.put_deploy.html) for more details
+#[no_mangle]
+pub extern "C" fn casper_put_deploy(
+    maybe_rpc_id: *const c_char,
+    node_address: *const c_char,
+    verbose: bool,
+    deploy_params: *const casper_deploy_params_t,
+    session_params: *const casper_session_params_t,
+    payment_params: *const casper_payment_params_t,
+    response_buf: *mut c_uchar,
+    response_buf_len: usize,
+) -> bool {
```

Comment:
I think @Fraser999 is right that using an `int` here would allow for success with `0` and an error code on failure.  The convention I have seen is to use negative numbers for error codes.

Here's how the `pigpio` library does it.  That library is for working with the GPIO pins on the Raspberry Pi:
https://github.com/joan2937/pigpio/blob/bb352d2000d8a4f38bab430927a85b7af4f70a01/pigpiod_if.h#L2974-L2987

The provide error messages for their codes in their [`pigpio_error`](https://github.com/joan2937/pigpio/blob/93965f8e3bc991a7c90155fae9dffe93e945c981/pigpiod_if.c#L410-L443) function.

I see `libsodium` follows the same pattern. Here's their error enum for their Argon2 implementation: 

https://github.com/jedisct1/libsodium/blob/0cabff7a0af454a5ea926182558099ff3f48f118/src/libsodium/crypto_pwhash/argon2/argon2.h#L76-L134